storage/remote/codec.go

```diff
@@ -183,6 +183,148 @@ func FromQueryResult(res *prompb.QueryResult) storage.SeriesSet {
 	}
 }
 
+func StreamChunkedReadResponses(
+	stream io.Writer,
+	queryIndex int64,
+	ss storage.SeriesSet,
+	sortedExternalLabels []prompb.Label,
+	sampleLimit int) error {
+	var numSamples = 0
+
+	for ss.Next() {
+		series := ss.At()
+		iter := series.Iterator()
+		lbls := MergeLabels(labelsToLabelsProto(series.Labels()), sortedExternalLabels)
+
+		for iter.Next() {
+			// TODO(bwplotka): Expose maxChunks as flag?
+			chks, samples, err := encodeChunks(iter, 120, 100, sampleLimit-numSamples)
+			if err != nil {
+				return err
+			}
+			numSamples += samples
+			if len(chks) == 0 {
+				break
+			}
+
+			b, err := proto.Marshal(&prompb.ChunkedReadResponse{
+				// TODO(bwplotka): Do we really need multiple?
+				ChunkedSeries: []*prompb.ChunkedSeries{
+					{
+						// TODO(bwplotka): Oneof? We don't need to send labels all the time.
+						Labels: lbls,
+						Chunks: chks,
+					},
+				},
+				QueryIndex: queryIndex,
+			})
+			if err != nil {
+				return errors.Wrap(err, "marshal ChunkedReadResponse")
+			}
+
+			if _, err := stream.Write(b); err != nil {
+				return errors.Wrap(err, "write to stream")
+			}
+		}
+
+		if err := iter.Err(); err != nil {
+			return err
+		}
+	}
+	if err := ss.Err(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// encodeChunks expects iterator to be ready to use (aka iter.Next() done before invoking).
+func encodeChunks(iter storage.SeriesIterator, maxSamplesPerChunk int, maxChunks int, sampleLimit int) ([]prompb.Chunk, int, error) {
+	var (
+		chkMint int64
+		chk     *chunkenc.XORChunk
+		app     chunkenc.Appender
+		err     error
+
+		isNext     = true
+		numSamples = 0
+		chks       = make([]prompb.Chunk, 0, maxChunks)
+	)
+
+	for isNext {
```

Comment:
I think you can do a more standard `for it := it.Next() {` here

Reply:
In current version I cannot due to:

> // encodeChunks expects iterator to be ready to use (aka iter.Next() done before invoking).

This simplifies caller code for us, but happy to rearrange it a bit.

Reply:
I'm not seeing that

Reply:
Hard to show you the issue, but this was the most clear implementation to me. Alternatives:

a)   passing `stream io.Write` into encodeChunks directly and stream straight away once chunks built up.

b)  duplicating `cut` process so it will end up like this:
```
// encodeChunks expects iterator to be ready to use (aka iter.Next() done before invoking).
func encodeChunks(iter storage.SeriesIterator, maxSamplesPerChunk int, maxChunks int, allowedNumSamples int) ([]prompb.Chunk, int, error) {
	var (
		chkMint int64
		chkMaxt int64
		chk     *chunkenc.XORChunk
		app     chunkenc.Appender
		err     error

		numSamples = 0
		chks       = make([]prompb.Chunk, 0, maxChunks)
	)

	for iter.Next() {
		numSamples++
		if allowedNumSamples > 0 && numSamples > allowedNumSamples {
			return nil, numSamples, HTTPError{
				msg:    fmt.Sprintf("exceeded sample limit"),
				status: http.StatusBadRequest,
			}
		}

		if chk == nil {
			chk = chunkenc.NewXORChunk()
			app, err = chk.Appender()
			if err != nil {
				return nil, numSamples, err
			}
			chkMint, _ = iter.At()
		}

		app.Append(iter.At())
		chkMaxt, _ = iter.At()

		if chk.NumSamples() < maxSamplesPerChunk {
			continue
		}

		// Cut the chunk.
		chks = append(chks, prompb.Chunk{
			MinTimeMs: chkMint,
			MaxTimeMs: chkMaxt,
			Type:      prompb.Chunk_Encoding(chk.Encoding()),
			Data:      chk.Bytes(),
		})
		chk = nil

		if maxChunks >= len(chks) {
			break
		}
	}
	if iter.Err() != nil {
		return nil, numSamples, errors.Wrap(iter.Err(), "iter TSDB series")
	}

	if chk != nil {
		// Cut the chunk if exists.
		chks = append(chks, prompb.Chunk{
			MinTimeMs: chkMint,
			MaxTimeMs: chkMaxt,
			Type:      prompb.Chunk_Encoding(chk.Encoding()),
			Data:      chk.Bytes(),
		})
	}
	return chks, numSamples, nil
}
```

Reply:
will try b let's see if we like it more